File: app/src/main/java/com/erikzuo/portfolioandroidapp/ui/main/MainActivity.java
package com.erikzuo.portfolioandroidapp.ui.main;

import android.arch.lifecycle.ViewModelProvider;
import android.arch.lifecycle.ViewModelProviders;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.view.GravityCompat;
import android.support.v7.app.ActionBarDrawerToggle;
import android.view.MenuItem;
import android.view.View;

import com.erikzuo.portfolioandroidapp.BR;
import com.erikzuo.portfolioandroidapp.R;
import com.erikzuo.portfolioandroidapp.databinding.ActivityMainBinding;
import com.erikzuo.portfolioandroidapp.databinding.NavHeaderMainBinding;
import com.erikzuo.portfolioandroidapp.ui.base.BaseActivity;
import com.erikzuo.portfolioandroidapp.ui.main.contact.ContactFragment;
import com.erikzuo.portfolioandroidapp.ui.main.home.HomeFragment;

import javax.inject.Inject;

import dagger.android.AndroidInjector;
import dagger.android.DispatchingAndroidInjector;
import dagger.android.support.HasSupportFragmentInjector;

/**
 * Created by YifanZuo on 3/2/18.
 */

public class MainActivity extends BaseActivity<ActivityMainBinding, MainViewModel> implements MainNavigator, HasSupportFragmentInjector {

    @Inject
    ViewModelProvider.Factory mViewModelFactory;

    @Inject
    DispatchingAndroidInjector<Fragment> fragmentDispatchingAndroidInjector;

    private MainViewModel mViewModel;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }


    @Override
    public MainViewModel getViewModel() {
        mViewModel = ViewModelProviders.of(this, mViewModelFactory).get(MainViewModel.class);

        return mViewModel;
    }

    @Override
    public int getBindingVariable() {
        return BR.viewModel;
    }

    @Override
    public int getLayoutId() {
        return R.layout.activity_main;
    }

    @Override
    public AndroidInjector<Fragment> supportFragmentInjector() {
        return fragmentDispatchingAndroidInjector;
    }

    @Override
    public void initViews() {
        setupNavDrawer();
        setupNavMenu();

        // When we first opened this activity, display Home fragment
        showFragment(
                HomeFragment.newInstance(),
                getString(R.string.home));
    }

    private void setupNavDrawer() {
        ActionBarDrawerToggle drawerToggle = new ActionBarDrawerToggle(
                this,
                getViewDataBinding().drawerView,
                getViewDataBinding().toolbar,
                R.string.open_drawer,
                R.string.close_drawer) {
            @Override
            public void onDrawerOpened(View drawerView) {
                super.onDrawerOpened(drawerView);
                hideKeyboard();
            }

            @Override
            public void onDrawerClosed(View drawerView) {
                super.onDrawerClosed(drawerView);
            }
        };

        getViewDataBinding().drawerView.addDrawerListener(drawerToggle);
        drawerToggle.syncState();
    }

    private void setupNavMenu() {
        NavHeaderMainBinding navHeaderMainBinding = DataBindingUtil.inflate(getLayoutInflater(),
                R.layout.nav_header_main, getViewDataBinding().navigationView, false);
        getViewDataBinding().navigationView.addHeaderView(navHeaderMainBinding.getRoot());
        navHeaderMainBinding.setViewModel(mViewModel);

        getViewDataBinding().navigationView.setNavigationItemSelectedListener(
                new NavigationView.OnNavigationItemSelectedListener() {
                    @Override
                    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
                        getViewDataBinding().drawerView.closeDrawer(GravityCompat.START);

                        mViewModel.setIsLoading(false);
                        switch (item.getItemId()) {
                            case R.id.navItemHome:
                                showFragment(
                                        HomeFragment.newInstance(),
                                        getString(R.string.home));

                                return true;
                            case R.id.navItemContact:
                                showFragment(
                                        ContactFragment.newInstance(),
                                        getString(R.string.contact));

                                return true;
                            default:
                                return false;
                        }
                    }
                });
    }

    private void showFragment(Fragment fragment, String title) {
        getSupportFragmentManager()
                .beginTransaction()
                .disallowAddToBackStack()
                .replace(R.id.clRootView, fragment)
                .commit();

        mViewModel.setPageTitle(title);
    }
}

File: app/src/main/java/com/erikzuo/portfolioandroidapp/ui/main/MainActivity.kt
package com.erikzuo.portfolioandroidapp.ui.main

import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.databinding.DataBindingUtil
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentTransaction
import android.support.v4.view.GravityCompat
import android.support.v7.app.ActionBarDrawerToggle
import android.view.View

import com.erikzuo.portfolioandroidapp.BR
import com.erikzuo.portfolioandroidapp.R
import com.erikzuo.portfolioandroidapp.databinding.ActivityMainBinding
import com.erikzuo.portfolioandroidapp.databinding.NavHeaderMainBinding
import com.erikzuo.portfolioandroidapp.ui.base.BaseActivity
import com.erikzuo.portfolioandroidapp.ui.main.education.EducationFragment
import com.erikzuo.portfolioandroidapp.ui.main.home.HomeFragment
import com.erikzuo.portfolioandroidapp.ui.main.work.WorkFragment
import com.erikzuo.portfolioandroidapp.ui.setting.SettingsActivity
import com.erikzuo.portfolioandroidapp.viewmodel.ViewModelProviderFactory

import dagger.android.AndroidInjector
import dagger.android.DispatchingAndroidInjector
import dagger.android.support.HasSupportFragmentInjector
import javax.inject.Inject

/**
 * Created by YifanZuo on 3/2/18.
 */

class MainActivity : BaseActivity<ActivityMainBinding, MainViewModel>(), MainNavigator, HasSupportFragmentInjector {


    @Inject
    internal lateinit var mFactory: ViewModelProviderFactory

    @Inject
    internal lateinit var fragmentDispatchingAndroidInjector: DispatchingAndroidInjector<Fragment>

    private lateinit var mViewModel: MainViewModel

    // Fragments
    private val mHomeFragment: HomeFragment = HomeFragment.newInstance()
    private val mWorkFragment: WorkFragment = WorkFragment.newInstance()
    private val mEducationFragment: EducationFragment = EducationFragment.newInstance()


    override val bindingVariable: Int
        get() = BR.viewModel

    override val layoutId: Int
        get() = R.layout.activity_main

    override val viewModel: MainViewModel
        get() {
            mViewModel = ViewModelProviders.of(this, mFactory).get(MainViewModel::class.java)
            return mViewModel
        }

    override fun supportFragmentInjector(): AndroidInjector<Fragment>? {
        return fragmentDispatchingAndroidInjector
    }


    override fun initViews() {
        viewModel.navigator = this
        setupNavDrawer()
        setupNavMenu()

        // When we first opened this activity, display Home fragment
        showFragment(
                mHomeFragment,
                getString(R.string.home))
    }

    private fun setupNavDrawer() {
        val drawerToggle = object : ActionBarDrawerToggle(
                this,
                binding.drawerView,
                binding.toolbar,
                R.string.open_drawer,
                R.string.close_drawer) {
            override fun onDrawerOpened(drawerView: View) {
                super.onDrawerOpened(drawerView)
                hideKeyboard()
            }
        }

        binding.drawerView.addDrawerListener(drawerToggle)
        drawerToggle.syncState()
    }

    private fun setupNavMenu() {
        val navHeaderMainBinding = DataBindingUtil.inflate<NavHeaderMainBinding>(
                layoutInflater,
                R.layout.nav_header_main,
                binding.navigationView,
                false)

        binding.navigationView.addHeaderView(navHeaderMainBinding.root)
        navHeaderMainBinding.viewModel = this.viewModel

        binding.navigationView.setNavigationItemSelectedListener { item ->
            binding.drawerView.closeDrawer(GravityCompat.START)

            this.viewModel.setIsLoading(false)
            when (item.itemId) {
                R.id.navItemHome -> {
                    showFragment(
                            mHomeFragment,
                            getString(R.string.home))

                    true
                }
                R.id.navItemWork -> {
                    showFragment(
                            mWorkFragment,
                            getString(R.string.work))

                    true
                }
                R.id.navItemEducation -> {
                    showFragment(
                            mEducationFragment,
                            getString(R.string.education))

                    true
                }
                R.id.navItemSetting -> {
                    goToSettings()

                    true
                }
                else -> false
            }
        }
    }

    private fun showFragment(fragment: Fragment, title: String) {
        val fragmentTransaction: FragmentTransaction = supportFragmentManager.beginTransaction()

        if (supportFragmentManager.findFragmentByTag(fragment.javaClass.name) == null) {
            fragmentTransaction
                    .addToBackStack( fragment.javaClass.name)
                    .add(R.id.clRootView, fragment, fragment.javaClass.name)
        }

        fragmentTransaction
                .hide(mHomeFragment)
                .hide(mWorkFragment)
                .hide(mEducationFragment)
                .show(fragment)
                .commit()

        this.viewModel.setPageTitle(title)
    }

    private fun goToSettings() {
        startActivity(Intent(this, SettingsActivity::class.java))
    }
}
